File: CPP/3-Behavioral/Null Object/NullObject.cpp
#include <iostream>
#include <vector>

class Item
{
public:
    virtual ~Item() {}
    virtual void do_something() = 0;
};

class FileItem : public Item
{
public:
    /*virtual*/ void do_something()
    {
        std::cout << "calculate checksum..." << std::endl;
    }
};

class NullItem : public Item
{
public:
    /*virtual*/ void do_something()
    {
    }
};


int main()
{
    std::vector<Item *> items;
    items.push_back(new FileItem);
    items.push_back(new NullItem);

    std::cout << "Item 0 (Active): "; items[0]->do_something();
    std::cout << "Item 1 (Null): "; items[1]->do_something();
    std::cout << std::endl;

    return 0;
}

//------------------------------------------------------------------------------
/*
Output:

Item 0: calculate checksum...
Item 1:

*/

File: Python/3-Behavioural/NullObject/NullObject.py
#!/usr/bin/env python3

from abc import ABC, abstractmethod


class Item(ABC):
    """
    It defines the common API of the NullObject and a valid Item object
    """
    def __init__(self):
        pass

    @abstractmethod
    def do_something(self):
        pass


class FileItem(Item):
    def do_something(self):
        return "Calculate file's checksum..."


class NullItem(Item):
    def do_something(self):
        pass


def main():
    items = [FileItem(), NullItem()]

    print("Item 0 (Active): %s" % items[0].do_something())
    print("Item 1 (Null): %s" % items[1].do_something())


if __name__ == '__main__':
    main()


"""
Output:

Item 0 (Active): Calculate file's checksum...
Item 1 (Null): None
"""
